tests/unit/morphing/test_concrete_provider.py

```diff
@@ -10,34 +10,68 @@
 import pytest
 from tests_helpers import cond_list, raises_exc
 
-from adaptix import Retort
+from adaptix import Provider, Retort
 from adaptix._internal.feature_requirement import HAS_PY_311, IS_PYPY
-from adaptix._internal.morphing.concrete_provider import DatetimeFormatProvider
+from adaptix._internal.morphing.concrete_provider import (
+    DatetimeFormatProvider,
+    DateTimestampProvider,
+    DatetimeTimestampProvider,
+)
 from adaptix.load_error import FormatMismatchLoadError, TypeLoadError, ValueLoadError
 
 
-def check_any_dt(loader):
-    raises_exc(
-        TypeLoadError(str, None),
-        lambda: loader(None),
-    )
-    raises_exc(
-        TypeLoadError(str, 10),
-        lambda: loader(10),
-    )
-    raises_exc(
-        TypeLoadError(str, datetime(2011, 11, 4, 0, 0)),
-        lambda: loader(datetime(2011, 11, 4, 0, 0)),
-    )
-    raises_exc(
-        TypeLoadError(str, date(2019, 12, 4)),
-        lambda: loader(date(2019, 12, 4)),
-    )
-    raises_exc(
-        TypeLoadError(str, time(4, 23, 1)),
-        lambda: loader(time(4, 23, 1)),
+@pytest.mark.parametrize(
+    ["expected_type", "invalid_objects", "loader_type", "extra_providers"],
```

Comment:
``loader_type`` sounds like "type of loader". You can use name ``tp`` to show which type is tested